### PR TITLE
Parse gcov 9.1 and 10 output and add tests for gcc 10

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: src/gcov.rs [hunk 1]
```diff
@@ -64,50 +64,92 @@ pub fn run_gcov(
...
+            parse_version(&output)
+        };
+    }
+    &V
+}
+
+pub fn get_gcov_output_ext() -> &'static str {
+    lazy_static! {
+        static ref E: &'static str = {
+            if get_gcov_version().major >= 10 {
```
Comment (marco-c) [inline src/gcov.rs:85]: Wasn't it introduced in gcc 9?

File: src/gcov.rs [hunk 2]
```diff
@@ -64,50 +64,92 @@ pub fn run_gcov(
...
+    let mut versions: Vec<_> = gcov_output
+        .split(|c| c == ' ' || c == '\n')
+        .filter_map(|value| Version::parse(value).ok())
+        .collect();
+    assert!(versions.len() >= 1, "no version found for `gcov`.");
+
+    versions.pop().unwrap()
+}
+
+pub fn check_gcov_version() -> bool {
```
Comment (marco-c) [inline src/gcov.rs:105]: Where is this one used other than tests? I couldn't find it

File: src/parser.rs [hunk 3]
```diff
@@ -313,6 +315,51 @@ pub fn parse_lcov(
     Ok(results)
 }
 
+#[derive(Debug, Default, Deserialize)]
+struct GcovJson {
+    format_version: String,
+    gcc_version: String,
+    current_working_directory: Option<String>,
+    data_file: String,
```
Comment (marco-c) [inline src/parser.rs:323]: What are `current_working_directory` and `data_file` used for?

File: src/parser.rs [hunk 4]
```diff
@@ -324,6 +371,48 @@ pub fn parse_gcov(gcov_path: &Path) -> Result<Vec<(String, CovResult)>, ParserEr
         .unwrap_or_else(|_| panic!("Failed to open gcov file {}", gcov_path.display()));
 
     let mut file = BufReader::new(&f);
+    if gcov_path.extension().unwrap() == "gz" {
```
Comment (marco-c) [inline src/parser.rs:374]: Let's split the function in two, one for `.json.gz` and one for `gcov`, then check the extension and call the right function, and return an error if the extension is unexpected

File: src/parser.rs [hunk 5]
```diff
@@ -324,6 +371,48 @@ pub fn parse_gcov(gcov_path: &Path) -> Result<Vec<(String, CovResult)>, ParserEr
...
+        let gz = GzDecoder::new(file);
+        let mut gcov: GcovJson = serde_json::from_reader(gz).unwrap();
```
Comment (marco-c) [inline src/parser.rs:376]: Maybe we should print a warning if the `format_version` is not the one we expect

File: src/gcov.rs (hunk 2 above)
Comment (calixteman) [reply src/gcov.rs:105]: I've no idea, I let this code because it was there...

File: src/parser.rs (hunk 3 above)
Comment (calixteman) [reply src/parser.rs:323]: `current_working_directory` is the output of `getpwd()` when the gcno is generated:
https://github.com/gcc-mirror/gcc/blob/e87559d202d90e614315203f38f9aa2f5881d36e/gcc/coverage.c#L1321
`data_file` is the gave as argument of `gcov`

File: src/gcov.rs (hunk 2 above)
Comment (marco-c) [reply src/gcov.rs:112]: Let's file an issue to remove it then

File: src/gcov.rs (hunk 1 above)
Comment (calixteman) [reply src/gcov.rs]: yep correct

File: src/lib.rs [hunk 6]
```diff
@@ -198,7 +202,14 @@ pub fn consumer(
                         }
 
                         let mut new_results = if gcov_type == GcovType::SingleFile {
-                            let new_results = try_parse!(parse_gcov(&gcov_path), work_item.name);
+                            let new_results = try_parse!(
+                                if gcov_ext.ends_with("gz") {
+                                    parse_gcov_gz(&gcov_path)
+                                } else {
```
Comment (marco-c) [inline src/lib.rs:208]: else endswith("gcov")
parse_gcov
else
log error

File: src/producer.rs [hunk 7]
```diff
@@ -118,7 +118,8 @@ impl Archive {
     fn is_gcno_llvm(reader: &mut dyn Read) -> bool {
         let mut bytes: [u8; 8] = [0; 8];
         reader.read_exact(&mut bytes).is_ok()
-            && bytes == [b'o', b'n', b'c', b'g', b'*', b'2', b'0', b'4']
+            && &bytes[..5] == b"oncg*"
+            && (&bytes[5..] == b"204" || &bytes[5..] == b"804")
```
Comment (marco-c) [inline src/producer.rs:122]: Why this change?

File: src/parser.rs (hunk 3 above)
Comment (marco-c) [reply src/parser.rs:323]: Can you add this as a comment to explain the fields?

File: src/producer.rs (hunk 7 above)
Comment (calixteman) [reply src/producer.rs:122]: to have integration test green with clang 11

File: src/parser.rs [hunk 8]
```diff
@@ -313,6 +315,105 @@ pub fn parse_lcov(
...
+    let f = File::open(&gcov_path)
+        .unwrap_or_else(|_| panic!("Failed to open gcov file {}", gcov_path.display()));
+
+    let file = BufReader::new(&f);
+    let gz = GzDecoder::new(file);
+    let mut gcov: GcovJson = serde_json::from_reader(gz).unwrap();
+    let mut results = Vec::new();
+
+    if gcov.format_version != "1" {
+        eprintln!(
```
Comment (marco-c) [inline src/parser.rs:373]: We usually use `log::error` for printing errors

**Review by marco-c — APPROVED** [review]

File: src/lib.rs (hunk 6 above)
Comment (calixteman) [reply src/lib.rs:208]: the test is made on `gcov_ext` set to either "gcov.json.gz" or "gcov" so there is no other alternative.

Comment (marco-c) [reply src/lib.rs:208]: That's true now, but who knows how the code will evolve, the check would allow us to be more defensive.
It's a nit though, feel free to ignore ;)

**Review by marco-c — APPROVED** [review]
LGTM!

Don't forget about filing the issue for removing check_gcov_version

**Review by marco-c — APPROVED** [review]

**Review by marco-c — APPROVED** [review]
